packages/shared/src/hooks/streaks/useStreakMilestone.ts

```diff
@@ -0,0 +1,55 @@
+import { useContext, useEffect } from 'react';
+import { useLazyModal } from '../useLazyModal';
+import { LazyModal } from '../../components/modals/common/types';
+import AlertContext from '../../contexts/AlertContext';
+import { useSettingsContext } from '../../contexts/SettingsContext';
+import { useReadingStreak } from './useReadingStreak';
+import { StreakModalProps } from '../../components/modals/streaks/common';
+
+export const useStreakMilestone = (): void => {
+  const { loadedSettings, optOutWeeklyGoal } = useSettingsContext();
+  const { streak, isLoading, isEnabled } = useReadingStreak();
+
+  const { loadedAlerts, alerts, updateAlerts } = useContext(AlertContext);
+  const { openModal, closeModal } = useLazyModal();
+
+  const modalType =
+    streak?.total === 1 ? LazyModal.FirstStreak : LazyModal.NewStreak;
+
+  useEffect(() => {
+    if (
+      !loadedAlerts ||
+      alerts?.showStreakMilestone !== true ||
+      isLoading ||
+      !isEnabled ||
+      !loadedSettings ||
+      optOutWeeklyGoal ||
+      !streak?.current
```

Comment:
Are the ors really what we want?
Feels like some are or and some should be and no?

It's a bit hard to follow due to the big evaluation.

@capJavert Possibly would like your review on this hook as well :)

Reply:
Alternative was thinking if we could simply put the high level exits (as optOut above the useEffect even?

Reply:
yeah, basically two things:

1. Minimize amount of dependencies by precomputing conditions outside of hook. Eg. instead of doing:
```js
useEffect(() => {
  if (cond1 && cond2 && !cond3 && cond4) {
    return
  }

  // logic
}, [cond1, cond2, cond3, cond4]);
```

do this:
```js
const condition = cond1 && cond2 && !cond3 && cond4

useEffect(() => {
  if (!condition) {
    return
  }

  // logic
}, [condition]);
```
`condition` variable name should reflect what the condition means. Also it lowers amount of dependencies that React has to check thus lowering number of `useEffect` executions.

2. As Chris said, the condition sometimes might be complicated for single evaluation and few early returns make it more readable so then it would look something like:
```js
const condition1 = cond1 && cond2

const condition2 = !cond3 && cond4

useEffect(() => {
  if (!condition1) {
    return
  }

  if (!condition2) {
    return
  }

  // logic
}, [condition1, condition2]);
```

Point 2. can be optional here but you can probably decide better since you are more in line on what code needs to do 👍

Reply:
We cannot put opt outs above hooks, because all hoooks need to be evaluated, they cannot be conditionally skipped: https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

Reply:
Please take a look again. I removed unneeded dependencies in the condition, split it into a separate variable and rewritten the code into multiple lines with comments.

Reply:
From my point of view good idea with `||=`. I would only check if we polyfill it correctly. The adoption is good but since some browsers got the support pretty recently https://caniuse.com/?search=Logical%20OR%20assignment I would want to be sure. Maybe check in compiled bundle how it looks?

Reply:
Oh, right. TIL that browsers didn't support that assignment before 😊 

I think I'll change to assigning directly. It's just syntactic sugar...